includes/enrolment/class-sensei-course-enrolment-manager.php

```diff
@@ -178,6 +186,74 @@ public function maybe_prevent_frontend_manual_enrol( $can_user_manually_enrol, $
 		return $can_user_manually_enrol;
 	}
 
+	/**
+	 * Run the deferred enrolment checks.
+	 *
+	 * @access private
+	 */
+	public function run_deferred_course_enrolment_checks() {
+		foreach ( $this->deferred_enrolment_checks as $user_id => $course_ids ) {
+			foreach ( array_keys( $course_ids ) as $course_id ) {
+				$this->do_course_enrolment_check( $user_id, $course_id );
+			}
+		}
+	}
+
+	/**
+	 * Defer course enrolment check to the end of request.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function defer_course_enrolment_check( $user_id, $course_id ) {
+		if ( ! isset( $this->deferred_enrolment_checks[ $user_id ] ) ) {
+			$this->deferred_enrolment_checks[ $user_id ] = [];
+		}
+
+		// Check if the enrolment check is already deferred.
+		if ( isset( $this->deferred_enrolment_checks[ $user_id ][ $course_id ] ) ) {
+			return;
+		}
+
+		// Usually the user will be back calculated by the end of the request, but mark them
+		// as needing a recalculation just in case the request fails early.
+		$this->mark_user_as_needing_recalculation( $user_id );
```

Comment:
This means that the next time we run the async job (we need to modify it slightly to periodically look for users that need recalculation) _or_ the _My Courses_ page loads, it will double check that we the user is indeed up-to-date, even if we ended up calculating this at the end of the request.